### PR TITLE
Fix ARM pcie-topology

### PR DESCRIPTION
**Pull request by ldoktor**
The  #2387 broke arm64-pci because the multifunction rootport was not defined for it's machine type. This PR fixes the definition and also adds some fixes in related code. I tested this on arm as well as i440fx and q35.

### CODE REVIEW

File: virttest/qemu_vm.py [hunk 1]
```diff
@@ -2373,31 +2373,30 @@ def sort_key(dev):
...
-                    else:
-                        port_addr = '%s.%s' % (func_0_addr, hex(func_num))
-                        root_port.set_param('addr', port_addr)
-                        devices.insert(root_port)
-                except DeviceError:
-                    logging.warning("No sufficient free slot for extra"
-                                    " root port, discarding %d of them"
-                                    % (extra_port_num - num))
-                    break
+        # Add extra root_port at the end of the command line only if there is
```
Comment (xutian) [inline virttest/qemu_vm.py]: @PaulYuuu , would you like to help review this part.  I'm not sure if ppc platform need the extra  pice root port.

File: virttest/qemu_devices/qdevices.py [hunk 2]
```diff
@@ -2236,7 +2236,8 @@ def is_direct_plug(self, device):
                         "vhost-vsock-pci", "virtio-net",
                         "virtio-scsi-pci", "virtio-balloon-pci",
                         "virtio-serial-pci", "virtio-rng-pci",
-                        "e1000e", "virtio-gpu-pci", "qemu-xhci"]
+                        "e1000e", "virtio-gpu-pci", "qemu-xhci",
+                        "VGA"]
```
Comment (luckyh) [inline virttest/qemu_devices/qdevices.py]: Seems different platforms have different topology representation, with using the following libvirt domain configuration,
```xml
<video>
  <model type='vga'/>
</video>
```
the outcome of qemu command is:
```
# x86_64 -M q35
    -device VGA,id=video0,vgamem_mb=16,bus=pcie.0,addr=0x1 \
```
```
# aarch64 -M virt
    -device pcie-root-port,port=0x8,chassis=1,id=pci.1,bus=pcie.0,multifunction=on,addr=0x1 \
    -device pcie-pci-bridge,id=pci.3,bus=pci.1,addr=0x0 \
    -device VGA,id=video0,vgamem_mb=16,bus=pci.3,addr=0x1 \
```
so, have to distinguish them, thanks.

**Comment by luckyh** [comment]
> ACK this change.

Sorry, left the message by mistake, please ignore it.

**Comment by luckyh** [comment]
@ldoktor ACK the rest part except https://github.com/avocado-framework/avocado-vt/pull/2420#discussion_r376222639, please help check it, thanks!

File: virttest/qemu_vm.py (hunk 1 above)
Comment (PaulYuuu) [reply virttest/qemu_vm.py]: AFAIK, pcc does not currently use pcie devices.

Comment (xutian) [reply virttest/qemu_vm.py]: @ldoktor  according @PaulYuuu 's comments, we need to ignore ppc platform.

Comment (ldoktor) [reply virttest/qemu_vm.py]: The "pcie_extra_root_port" is only defined for aarch64-pci and q35 so there is no need to explicitly filter this. Once/if ppc64/riscv starts using pcie and we ask for them, it should receive them without any modification to this code....

File: virttest/qemu_devices/qdevices.py (hunk 2 above)
Comment (ldoktor) [reply virttest/qemu_devices/qdevices.py]: I don't think it's worth changing, let's keep it the q35 way (original).

I assume there used to be a problem putting the VGA to a pcie-root-port before so they added the hook to inject it directly. On arm nobody probably tried that so it uses the default. I guess (don't know for sure) that they will start plugging the VGA to a root-port in the future and then we might add it to the list of direct-plugs.

So let me simply remove this patch for now :smile:

Comment (luckyh) [reply virttest/qemu_devices/qdevices.py]: ok, sounds good, thanks.

**Comment by ldoktor** [comment]
Changes:

* removed the commit that puts VGA behind root-port to comply to q35 default libvirt topology
* __not changed__ the extra pcie root ports as they are only enabled via configuration so ppc64le is not affected and it's better to leave it this way as once we do enable them on other archs they will become available via the same configuration option.

File: virttest/qemu_vm.py (hunk 1 above)
Comment (xutian) [reply virttest/qemu_vm.py]: @ldoktor  some tests configured  'pcie-extra_root_port'  but these case are not x86/x86_64 only tests.  so I suggest to keep original code block:

xutian$ grep 'pcie_extra_root_port'  -R |grep cfg
qemu/tests/cfg/smbios.cfg:    pcie_extra_root_port = 0
qemu/tests/cfg/block_hotplug_in_pause.cfg:                pcie_extra_root_port = 2
qemu/tests/cfg/virtio_console.cfg:                                pcie_extra_root_port = 4
qemu/tests/cfg/rng_hotplug.cfg:                pcie_extra_root_port = 4
qemu/tests/cfg/block_hotplug.cfg:                pcie_extra_root_port = 2
qemu/tests/cfg/seabios_hotplug_unplug.cfg:        pcie_extra_root_port = 2
qemu/tests/cfg/nic_hotplug.cfg:                pcie_extra_root_port = 2
qemu/tests/cfg/nic_hotplug.cfg:                    pcie_extra_root_port = 0
qemu/tests/cfg/multi_nics_hotplug.cfg:        pcie_extra_root_port = 4

Comment (ldoktor) [reply virttest/qemu_vm.py]: The original code does not work as there are other archs that require the pcie_extra_ports (for now just arm and q35, but this should expand with riscv in the future and perhaps pseries will start utilize pcie aswell...).

How about looking at busses and if the machine contains PCIE bus add the extra root ports?

**Comment by xutian** [comment]
ACK, double checked,  no impact for ppc platform. thanks @PaulYuuu  and @ldoktor

**Pull request merged**